kitsune/search/v2/base.py

```diff
@@ -126,3 +129,91 @@ def prepare_locale(self, instance):
         if getattr(instance, "locale"):
             return instance.locale
         return ""
+
+
+class SumoSearch(ABC):
+    """Base class for search classes.
+
+    Implements the run() function, which will perform the search.
+
+    Child classes should define values for the various abstract properties this
+    class has, relevant to the documents the child class is searching over.
+    """
+
+    def __init__(self, locale, product=None):
+        self.locale = locale
+        self.product = product
+        self.results_per_page = 10
+
+    def run(self, query, page=1):
+        """Perform search, placing the results in `self.results`, and the total
+        number of results (across all pages) in `self.total`. Chainable."""
+
+        search = DSLSearch(using=es7_client(), index=self.index)
+
+        # add the search class' filter
+        # `should` with `minimum_should_match=1` acts like an OR filter
+        search = search.query("bool", should=self.filter, minimum_should_match=1)
+
+        # add query, search over the search class' fields
+        search = search.query(
+            "simple_query_string", query=query, default_operator="AND", fields=self.fields
+        )
+
+        # add highlights for the search class' highlight_fields
+        search = search.highlight(
+            *self.highlight_fields,
+            type="fvh",
+            # order highlighted fragments by their relevance:
+            order="score",
+            # only get one fragment per field:
+            number_of_fragments=1,
+            # split fragments at the end of sentences:
+            boundary_scanner="sentence",
+            # return fragments roughly this size:
+            fragment_size=SNIPPET_LENGTH,
+            # add these tags before/after the highlighted sections:
+            pre_tags=[f"<{HIGHLIGHT_TAG}>"],
+            post_tags=[f"</{HIGHLIGHT_TAG}>"],
+        )
+
+        # do pagination
+        start = (page - 1) * self.results_per_page
+        search = search[start : start + self.results_per_page]
+
+        # perform search
+        self.hits = search.execute().hits
+
+        self.total = self.hits.total.value if self.hits else 0
+        self.results = [self.make_result(hit) for hit in self.hits]
+
+        return self
+
+    @property
```

Comment:
I am not sure about the `@property` in all the abstractmethods in here. Any particular reason you r going down this road?